### PR TITLE
[coap] reject option number overflow in Option::Iterator::Advance()

### PR DESCRIPTION
**Pull request by 5h4d0wr007**
Guards the cumulative option-number accumulation in `Option::Iterator::Advance()` with `CanAddSafely<uint16_t>` so a delta that would overflow past 0xFFFF is rejected with `kErrorParse` instead of wrapping. Regression test is also added.

Ref: https://github.com/openthread/openthread/issues/13578

### CODE REVIEW

**Review by abtink — APPROVED** [review]
Thanks. LGTM.

File: src/core/coap/coap_message.cpp [hunk 1]
```diff
@@ -829,6 +829,7 @@ Error Option::Iterator::Advance(void)
     VerifyOrExit(optionLength <= GetMessage().GetLength() - mNextOptionOffset, error = kErrorParse);
     mNextOptionOffset += optionLength;
 
+    VerifyOrExit(CanAddSafely<uint16_t>(mOption.mNumber, optionDelta), error = kErrorParse);
```
Comment (jwhui) [inline src/core/coap/coap_message.cpp:832]: `mNextOptionOffset += optionLength;` is executed on line 830 before this check. If `CanAddSafely` fails, mNextOptionOffset has already been incremented.

Furthermore, we can fail fast even earlier: `optionDelta` is already fully decoded right after `ReadExtendedOptionField(optionDelta)` on line 824. If we check `CanAddSafely` immediately after, this would avoid reading the extended length bytes from the message when the option delta is already known to overflow, and ensures no iterator state is mutated before all validations pass.

File: tests/unit/test_coap_overflow.cpp [hunk 2]
```diff
@@ -67,6 +67,47 @@ void TestCoapOverflow(void)
...
+    SuccessOrQuit(message->AppendBytes(options, sizeof(options)));
+
+    Coap::Option::Iterator iterator;
+
+    SuccessOrQuit(iterator.Init(*message));
+    VerifyOrQuit(!iterator.IsDone());
+    VerifyOrQuit(iterator.GetOption()->GetNumber() == 65535);
+
+    // The second option would wrap the running option number and must be rejected.
+    VerifyOrQuit(iterator.Advance() == kErrorParse,
```
Comment (jwhui) [inline tests/unit/test_coap_overflow.cpp:104]: Should also verifyt that the iterator is marked done and no longer returns an option pointer after the error.

File: tests/unit/test_coap_overflow.cpp [hunk 3]
```diff
@@ -67,6 +67,47 @@ void TestCoapOverflow(void)
...
+                 "Fix failed: cumulative option number overflow was not rejected!");
+
+    message->Free();
+    testFreeInstance(instance);
```
Comment (jwhui) [inline tests/unit/test_coap_overflow.cpp:108]: It may also be worth testing the boundary condition where `delta == 0` when `mOption.mNumber == 65535` to verify that valid repeated options at `0xffff` are still accepted.

File: src/core/coap/coap_message.cpp (hunk 1 above)
Comment (5h4d0wr007) [reply src/core/coap/coap_message.cpp]: Good point. Moved the check to right after `ReadExtendedOptionField(optionDelta)` so we bail before decoding the extended length field.

File: tests/unit/test_coap_overflow.cpp (hunk 2 above)
Comment (5h4d0wr007) [reply tests/unit/test_coap_overflow.cpp:104]: Added `VerifyOrQuit(iterator.IsDone())` and `VerifyOrQuit(iterator.GetOption() == nullptr)`
after the error.

File: tests/unit/test_coap_overflow.cpp (hunk 3 above)
Comment (5h4d0wr007) [reply tests/unit/test_coap_overflow.cpp:145]: Added `TestCoapOptionNumberMaxRepeated()` covering `delta == 0` at 65535. Verified it
fails if the check is tightened to `< 0xffff`, so the boundary is genuinely guarded.

File: tests/unit/test_coap_overflow.cpp [hunk 4]
```diff
@@ -67,6 +67,84 @@ void TestCoapOverflow(void)
...
+
+    SuccessOrQuit(iterator.Advance());
+    VerifyOrQuit(!iterator.IsDone());
+    VerifyOrQuit(iterator.GetOption()->GetNumber() == 65535);
+
+    SuccessOrQuit(iterator.Advance());
+    VerifyOrQuit(iterator.IsDone());
+
+    message->Free();
+    testFreeInstance(instance);
```
Comment (jwhui) [inline tests/unit/test_coap_overflow.cpp:145]: Add `VerifyOrQuit(iterator.GetOption() == nullptr);` here as well for consistency with `TestCoapOptionNumberOverflow()`?

File: tests/unit/test_coap_overflow.cpp [hunk 5]
```diff
@@ -67,6 +67,84 @@ void TestCoapOverflow(void)
...
+}
+
+void TestCoapOptionNumberMaxRepeated(void)
+{
+    Instance      *instance;
+    Coap::Message *message;
+
+    uint8_t options[] = {0xe0, 0xfe, 0xf2, 0x00};
```
Comment (jwhui) [inline tests/unit/test_coap_overflow.cpp:118]: Add a brief comment here (similar to the one in `TestCoapOptionNumberOverflow()`) describing what this byte sequence tests.

Also, make this static constexpr uint8_t kOptions[].

```suggestion
    // Option 1: delta 65535 (0xe0, 0xfe, 0xf2), length 0.
    // Option 2: delta 0 (0x00), length 0.
    // Verifies that repeated options with delta 0 at max option number 65535 are accepted.

    static constexpr uint8_t kOptions[] = {0xe0, 0xfe, 0xf2, 0x00};
```

File: tests/unit/test_coap_overflow.cpp [hunk 6]
```diff
@@ -67,6 +67,84 @@ void TestCoapOverflow(void)
...
+    // Two options with deltas 65535 then 12. Each delta is individually
+    // valid, but their cumulative sum (65547) exceeds the 16-bit option
+    // number and would wrap to 11 (Uri-Path), producing a non-monotonic
+    // option sequence that RFC 7252 forbids.
+    //
+    // Byte 0xe0 : option 1, delta 14 (2-byte extension), length 0.
+    // Bytes 0xfe 0xf2 : extended delta 0xfef2 + 269 = 65535.
+    // Byte 0xc0 : option 2, delta 12, length 0.
+
+    uint8_t options[] = {0xe0, 0xfe, 0xf2, 0xc0};
```
Comment (jwhui) [inline tests/unit/test_coap_overflow.cpp:84]:
```suggestion
    static constexpr uint8_t options[] = {0xe0, 0xfe, 0xf2, 0xc0};
```

**Comment by 5h4d0wr007** [comment]
Thanks for the review. I have addressed the comments, and squashed into a single commit.

**Review by jwhui — APPROVED** [review]
Thanks! 👍🏼
